DRILL-6348: Fixed code so that Unordered Receiver reports its memory …

sachouche opened the pull request:
Problem Description -
- The Unordered Receiver doesn't report any memory usage because the RPC infrastructure associated the allocated memory ownership to the minor fragment container

Proposed Fix -
- Modified the code to change the received DrillBuf memory ownership from the parent fragment to the operator (as discussed with @parthchandra)
- Made sure that the buffer accounting logic is preserved

@parthchandra, can you please review this pull request?

Thanks!

HanumathRao commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/unorderedreceiver/UnorderedReceiverBatch.java:
```diff
@@ -182,13 +184,18 @@ public IterOutcome next() {
       }
 
+      // Transfer the ownership of this raw-batch to this operator for proper memory statistics reporting
+      batch = batch.transferBodyOwnership(oContext.getAllocator());
```
@sachouche  Thank you for the changes. I was just contemplating if the below change
batch = batch.transferBodyOwnership(oContext.getAllocator()); 
needs to be moved before the check at line number 183.

sachouche replied:
@HanumathRao,

- This will not affect the outcome since the unordered-receiver is a child of the context allocator (the minor fragment)
- Doing the check earlier seemed cleaner as there is no point of changing ownership when there is already an out-of-memory condition
- I think the code was written this way since implicitly, the network handlers are receiving the batch and then performing a change of ownership (from RPC to context allocator); this step could lead to an out-of-memory condition at the fragment level
- If your point is about reporting, that is attributing the OOM condition to the child operator, then I guess I don't have a problem with that.
@parthchandra what do you think?

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/record/RawFragmentBatch.java:
```diff
@@ -77,4 +83,46 @@ public long getByteCount() {
+    // Clear the current Drillbuffer since caller will perform release() on the new one
+    body.release();
+
+    return new RawFragmentBatch(getHeader(), transferResult.buffer, getSender(), false);
```
Why is it necessary to return new `RawFragmentBatch` instead of setting `body` to `transferResult.buffer`?

vrozov replied:
I don't see where RawFragmentBatch is cached. Is not it removed from a queue using poll()?

sachouche replied:
Look at the IncomingBuffers.batchArrived:
- It holds a reference to the RawFragmentBatch
- Holds a lock on the parent collector for insertion
- Though, the collector enqueue method itself is not synchronized on the "this" object
- Instead, it relies on the synchronization provided as part of the RawBatchBuffer queue
- This means a getNext() call will be able to dequeue a RawFragmentBatch instance
- Concurrently, the IncomingBuffers.batchArrived will invoke release on the cached reference

vrozov replied:
This actually brings a question why `newRawFragmentBatch` is released in `IncomingBuffers.batchArrived()` instead of releasing `transferredBuffer` after `RawFragmentBatch` is constructed in `newRawFragmentBatch`.

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/record/RawFragmentBatch.java:
```diff
@@ -77,4 +83,46 @@ public long getByteCount() {
+    }
+
+    int writerIndex               = body.writerIndex();
+    TransferResult transferResult = body.transferOwnership(targetAllocator);
```
There is no check that the transfer is within new allocator limits, is it guaranteed to succeed?

vrozov replied:
But what if it is an over limit?

sachouche replied:
- After a call to getNext(), the operator code checks whether the transfer has succeeded
- So that an OOM condition is reported within the fragment
- When the change of ownership happens from parent to child, there should not be any OOM condition since this is internal account

For code clarity, I can improve the javadoc to ask the caller to check for an OOM condition (similarly to the BaseAlocator behavior).

vrozov commented on exec/memory/base/src/main/java/org/apache/drill/exec/memory/AllocationManager.java:
```diff
@@ -253,10 +261,12 @@ public boolean transferBalance(final BufferLedger target) {
+        // Release first to handle the case where the current and target allocators were part of the same
+        // parent / child tree.
         allocator.releaseBytes(size);
+        boolean allocationFit = target.allocator.forceAllocate(size);
```
move `owningLedger = target;` before `return target.allocator.forceAllocate(size)` as there is no check for the result of the call anyway.

vrozov replied:
If this happens, is not there a problem that the old allocator already released the memory? In any case, won't runtime exception cancel the query anyway and all allocators will be closed.

sachouche replied:
What about debugging:
- An operation fails
- And yet there is a change of ownership
 
@vrozov, I rather not make this change as I strongly believe the change of ownership should happen only on success.

vrozov replied:
In this case, changing the order of `forceAllocate()` and `releaseBytes()` is incorrect as ownership is not changed, but the old owner does not account for that memory anymore.

sachouche replied:
- The change of order is an optimization for a parent / child relationship as if we don't release first, then we could unnecessarily go over the memory budget (double counting).
- The force-alloc() / free() failures should never happen on normal conditions; when they do, the best thing to do is to exit. I still prefer not to promote the target allocator till it is 100% successful.

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/unorderedreceiver/UnorderedReceiverBatch.java:
```diff
@@ -153,8 +153,10 @@ private RawFragmentBatch getNextBatch() throws IOException {
+    RawFragmentBatch batch = null;
     try{
-      RawFragmentBatch batch;
+
```
Consider moving try block to a separate function.

vrozov replied:
create function `getNextNotEmptyBatch()` that calls `getNextBatch()` and either returns not empty batch or `null`.

sachouche replied:
will do. Thanks for the suggestion.

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/unorderedreceiver/UnorderedReceiverBatch.java:
```diff
@@ -182,13 +184,18 @@ public IterOutcome next() {
       stats.addLongStat(Metric.BYTES_RECEIVED, batch.getByteCount());
 
       batch.release();
+      batch = null;
```
Why can't it be done in finally?

sachouche replied:
@vrozov 

My bad, the highlight was on the "batch = null" code; I guess, you meant why can't we move the whole release logic within the finally block. I agree with your proposal as delaying a bit the release phase doesn't hurt u in this case. I will make the change.

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/unorderedreceiver/UnorderedReceiverBatch.java:
```diff
@@ -153,8 +153,10 @@ private RawFragmentBatch getNextBatch() throws IOException {
     batchLoader.resetRecordCount();
     stats.startProcessing();
+
+    RawFragmentBatch batch = null;
```
It is not obvious that `body` is `null` when the `if` condition is met. Additionally, what if a release logic changes in future? It is a bad practice to rely on the details of implementation.

sachouche commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/unorderedreceiver/UnorderedReceiverBatch.java:
```diff
@@ -149,25 +149,32 @@ private RawFragmentBatch getNextBatch() throws IOException {
     }
   }
 
+  private RawFragmentBatch getNextNotEmptyBatch() throws IOException {
```
why is that? this method is used once and doing so will make us duplicate exception handling code. I would appreciate if you provide reason(s) when you make suggestions to avoid the back and forth.

vrozov replied:
No, there is no need to handle `IOException` twice. There are no other methods that throw `IOException`. `SchemaChangeException` is actually never thrown. Please see *TODO* comment and DRILL-2933.

sachouche commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/unorderedreceiver/UnorderedReceiverBatch.java:
```diff
@@ -149,25 +149,32 @@ private RawFragmentBatch getNextBatch() throws IOException {
+  private RawFragmentBatch getNextNotEmptyBatch() throws IOException {
+    RawFragmentBatch batch;
+    try {
+      stats.startWait();
```
Sure, but does it really matter or is it your personal preference?

vrozov replied:
Yes, it does matter. If something goes wrong during `startWait()` there should be no call to `stopWait()`.

vrozov replied:
it may throw `AssertException` now and other exceptions may be added in the future.

sachouche replied:
I see; I will then fix any such occurrences when opportunity presents itself as I have seen both patterns in the Drill code base.

sachouche commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/unorderedreceiver/UnorderedReceiverBatch.java:
```diff
@@ -149,25 +149,32 @@ private RawFragmentBatch getNextBatch() throws IOException {
+      batch = getNextBatch();
+
+      // skip over empty batches. we do this since these are basically control messages.
+      while (batch != null && batch.getHeader().getDef().getRecordCount() == 0
```
Sure. 

FYI - We cannot release the batch within this method.

vrozov replied:
Please explain why it can't be released. Note that the release is done for the batch that will *not* be returned from the method.

sachouche commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/unorderedreceiver/UnorderedReceiverBatch.java:
```diff
@@ -201,6 +208,11 @@ public IterOutcome next() {
+
+      if (batch != null) {
+        batch.release();
+        batch = null;
```
This pattern ensures that after release the batch cannot be accessed anymore (e.g., additions to the current code).

vrozov replied:
OK, but note that additions to the current code can also delete `batch = null` assignment and that `batch` can be used after `release()` call. For example, with the current implementation, `batch.getHeader()` is perfectly valid after the batch was released.

sachouche replied:
The point of this pattern is that if you would like to continue using this object then be prepared to know what can and what cannot be used.

sachouche commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/unorderedreceiver/UnorderedReceiverBatch.java:
```diff
@@ -149,25 +149,32 @@ private RawFragmentBatch getNextBatch() throws IOException {
+      batch = getNextNotEmptyBatch();
       first = false;
 
       if (batch == null) {
```
Sure.

sachouche commented:
@vrozov, your observation is valid, we need more JIRAs to fix the reporting problem
**Current Fix**
- At this time, the UnorderedReceiver didn't account for any consumed memory
- This fix, taxes the operator only when it consumes buffers

**Potential Enhancements**
Solution I
- Create a new fragment child Allocator which will own the received (not yet consumed) batches
- Improve the UI to report this allocator size
- This solution is simple and complementary to the current work

Solution II
- Have the receiver operator drain the batch queue
- Essentially, the receiver will have a private queue from where to consume batches
- I personally don't like this solution as it prevents us from improving the Drill network protocol
- Draining the batches for the sake of reporting will make it harder for the network layer to prefetch batches in an optimal manner; the queue size should be an indicator on how many pending batches there are.

vrozov commented:
IMO, it is better not to report memory usage at all compared to reporting a wrong number. In case incoming batches are accumulated in a queue, they should be reported as owned by a receiver. Taking ownership just before passing a batch to the next operator does not sound right to me.

I don't think it is necessary to create new fragment child allocator. Receiver allocator should be used instead of fragment allocator when an incoming batch is placed into a queue.

sachouche commented:
@vrozov,

**What are we trying to solve / improve**
- Drill is currently not properly reporting memory held in Fragment's receive queues
- This makes it hard to analyze OOM conditions
This is what I want to see
- Every operator reporting on the resources it is currently using (needed)
- Fragment held resources (other than the ones already reported by the child operators)
- Drilbit level (metadata caches, web-server, ..)
- I am ok to incrementally reach this goal

**Data Exchange Logistic**
- Ideally, the data exchange fabric should be decoupled from the Drill Receive / Send operators
- The fabric should be handling all the aspects of pre-fetch / pressuring and so forth
- It will tune to the speed of producers / consumers when writing / reading data from it
- This infrastructure should have its own resource management and reporting capabilities

**Operator based Reporting**
- Receive and Send operators shall not worry about batches they didn't consume yet
- Doing so is counter productive as the Data Exchange fabric will interpret a "drain" operation as the operator "needing" more data. 
- For example, the merge-receiver should not be managing the receive queues; it should only advertise the pattern of data consumption and let the exchange fabric figure out the rest. 

The main difference in the two approaches, is that essentially, you are preaching for Receive and Send operators to control all aspects of communication whereas I am preaching for decoupling such aspects.

vrozov commented:
@sachouche I'd suggest moving the discussion to dev list as the topic of the batch ownership is beyond PR review (code changes).

sachouche commented:
That was not my intention as my current change aimed at describing the system the way it is. 

@parthchandra, any feedback?

vrozov commented:
IMO, it will be good to understand what other operators do as well. For example what Project or Filter operators do. Do they take ownership of incoming batches? And if they do, when is the ownership taken?

I do not suggest that we change how Sender and Receiver control **all** aspects of communication, at least not as part of this JIRA/PR. The difference in my and your approach is whether or not UnorderedReceiver and other receivers are pass-through operators. My view is that receivers are not pass-through operators and they are buffering operators as they receive batches from the network and buffer them before downstream operators are ready to consume those batches. In your view, receivers are pass-through operators that get batches from fragment queue or some other queue and pass them to downstream. As there is no wait and no processing between getting a batch from fragment queue and passing it to the next operator, I don't see why a receiver needs to take the ownership.

parthchandra commented on exec/java-exec/src/main/java/org/apache/drill/exec/record/RawFragmentBatch.java:
```diff
@@ -77,4 +83,47 @@ public long getByteCount() {
+   *         been switched to the target allocator); otherwise this operation is a NOOP (current instance
+   *         returned)
+   */
+  public RawFragmentBatch transferBodyOwnership(BufferAllocator targetAllocator) {
```
DrillBuf has a transferOwnership method that could be used instead? See the transferTo methods in any of the vector classes -  [FixedValueVectors.transferTo(....)](https://github.com/apache/drill/blob/master/exec/vector/src/main/codegen/templates/FixedValueVectors.java#L282) 
Saves you the trouble of rolling your own and also avoids the need to make changes to any of the existing classes.

parthchandra commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/unorderedreceiver/UnorderedReceiverBatch.java:
```diff
@@ -182,13 +189,15 @@ public IterOutcome next() {
       }
 
+      // Transfer the ownership of this raw-batch to this operator for proper memory statistics reporting
+      batch = batch.transferBodyOwnership(oContext.getAllocator());
```
This should probably be done inside the [ RecordBatchLoader.load() ](https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/record/RecordBatchLoader.java#L78) method.
Note that `MergingRecordBatch` has similar code and so probably suffers from the same memory accounting issue. All other uses of `RecordBatchLoader.load()` appear to be in the client code or test code so we are unlikely to break anything by making the change in `RecordBatchLoader`.

sachouche commented:
Met with @parthchandra  and  @vrozov to discuss a more comprehensive fix:

**Agreement**
It was agreed that received batches should be owned by the associated receiver (not the fragment).
This association is done at the framework level (Data Collector) so that the receiver doesn't have to perform any extra processing (such as explicit draining); this is to ensure that no side effect will occur (e.g., acknowledgment logic since it is sensitive to operator record consumption)

**Fix**
- Modified the Unordered & Merge receivers to register their buffer allocators with the associated Data Collector
- The IncomingBuffers class now uses the operator's buffer allocator instead of the fragment allocator

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/MergingReceiverCreator.java:
```diff
@@ -44,6 +44,11 @@ public MergingRecordBatch getBatch(ExecutorFragmentContext context,
+    MergingRecordBatch mergeReceiver = new MergingRecordBatch(context, receiver, buffers);
+
+    // Register this operator's buffer allocator so that incoming buffers are owned by this allocator
+    bufHolder.setOprAllocator(receiver.getOppositeMajorFragmentId(), mergeReceiver.getOprAllocator());
```
Consider moving registration of the buffer allocator inside `MerginRecordBatch` constructor (change constructor to accept `ExchangeFragmentContext` and `MergingReceiverPOP` only).

sachouche replied:
sure.

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/unorderedreceiver/UnorderedReceiverCreator.java:
```diff
@@ -40,6 +40,11 @@ public UnorderedReceiverBatch getBatch(ExecutorFragmentContext context, Unordere
     assert buffers.length == 1;
     RawBatchBuffer buffer = buffers[0];
-    return new UnorderedReceiverBatch(context, buffer, receiver);
+    UnorderedReceiverBatch receiverBatch = new UnorderedReceiverBatch(context, buffer, receiver);
```
The same as for `MerginRecordBatch`.

sachouche replied:
sure.

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/batch/IncomingBuffers.java:
```diff
@@ -129,6 +134,10 @@ public int getRemainingRequired() {
     return collectorMap.get(senderMajorFragmentId).getBuffers();
   }
 
+  public void setOprAllocator(int senderMajorFragmentId, BufferAllocator oprAllocator) {
```
Consider introducing `getCollector(int senderMajorFragmentId)` instead of `setOprAllocator` and `getBuffers`;

sachouche replied:
I feel the Collector is an internal class (also used for synchronization). Why exposed it if this is not necessary?

vrozov replied:
`DataCollector` is a top-level interface.

sachouche replied:
I don't mind doing so if you think Collector is a top-level interface. I wish there is a better way to indicate which class is an internal framework class vs a public one safe for all operators to access. If you look closely at the code, the DataCollector interface was created to allow multiple implementations; the Drill code exposed the RawFragmentBatchProvider as the main IF for operators to consume incoming data.

vrozov replied:
It can be done using annotations (I do not suggest to introduce such annotations now).

I don't see a need to hide DataCollector as long as HashMap is not exposed directly or concrete type of a class that implements DataCollector is exposed.

sachouche commented:
Implemented the changes suggested by @vrozov and fixed a memory leak warning message as the receive operators were closed before the buffers. Now closing the buffers before closing the operators.

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/batch/DataCollector.java:
```diff
@@ -28,4 +29,16 @@
+   *
+   * @param oprAllocator operator buffer allocator
+   */
+  void setOprAllocator(BufferAllocator oprAllocator);
```
I am not sure that `Opr` qualification is necessary. There is no assumption that it is an operator. It can be any `BufferAllocator`.

sachouche replied:
Agreed. What about get/setOwningAllocator?

vrozov replied:
IMO, get/setAllocator is good enough.

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/batch/IncomingBuffers.java:
```diff
@@ -103,8 +104,12 @@ public boolean batchArrived(final IncomingDataBatch incomingBatch) throws Fragme
 
+      // Use the Data Collector's buffer allocator if set, otherwise the fragment's one
+      BufferAllocator ownerAllocator =
+        collector.getOprAllocator() != null ? collector.getOprAllocator() : context.getAllocator();
```
Consider moving this logic to `AbstractDataCollector.getAllocator()` or `AbstractDataCollector.setAllocator()`.

sachouche commented:
@vrozov, made the changes but note that I had to change the DataCollector IF visibility to public; as previously indicated, this IF was used only within the exec.work.batch package.

vrozov commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/batch/AbstractDataCollector.java:
```diff
@@ -116,4 +120,16 @@ public void close() throws Exception {
+  /** {@inheritDoc} */
+  @Override
+  public void setAllocator(BufferAllocator allocator) {
+    this.ownerAllocator = allocator;
```
Can passed allocator be `null`? If yes, `ownerAllocator` should not probably be overwritten. If no, please add an `assert` or better `Preconditions` check.

sachouche replied:
I wondered about the same; the only reasons for the caller to pass a null, is either a bug or allocator unregistration. I guess for now, I'll add an assert since we do not really need an unregistration step (this will also help us catch erroneous setting of the allocator).

vrozov replied:
I'd recommend `Preconditions` over `assert`.

sachouche replied:
sure.

arina-ielchiieva commented:
@sachouche travis build fails with check style errors. Please fix.

sachouche commented:
@arina-ielchiieva 
I have fixed the Travis issue.
